### ISSUE TITLE
2025.04.09 - #30

### ISSUE THREAD

**Issue by changh95** [changh95/WeeklySpatialAI#32]

**Comment by james-joobs** [comment]
> 📎 관련 링크: [llama.cpp의 grammar 문법 보기](https://github.com/ggml-org/llama.cpp/tree/master/grammars)

# 📦 확률적인 LLM에게 JSON을 "강요"하는 법  


**— Structured Output, Grammar, 그리고 마스킹의 원리**

LLM은 기본적으로 “확률적 텍스트 생성기”다.  
하지만 우리는 실제 서비스에서 그들에게 `"function_call"` 같은 JSON을 요구하고, 그걸 기계적으로 정확하게 출력하길 기대한다.  
**어떻게 그런 일이 가능할까?**

---

## 🔍 LLM은 본래 "확률적"이다

LLM은 한 번에 하나의 토큰(token)을 생성한다.  
이때 각 토큰은 그 시점의 문맥(context)을 기반으로 확률 분포를 가지며, 그 중 하나가 샘플링된다.

예:  
```
입력: 오늘 날씨는  
출력 후보: [맑 → 0.6, 흐 → 0.3, 비 → 0.1]  
```

이런 식으로 LLM은 다음 단어(토큰)를 선택해 나가며 문장을 완성한다.  
**즉, 생성 결과는 "확률적으로 움직이는 선택의 연속"이다.**

---

## ❗ 그럼에도 어떻게 함수 호출, JSON 같은 "형식적인 구조"를 정확히 생성할 수 있는가?

바로 여기서 등장하는 것이 **Structured Output** 기능이다.  
이 기능은 OpenAI의 `"function_calling"`, `"json_mode"`, Huggingface의 `structured output`, `grammar`, `jsonformer`, llama.cpp의 `GBNF grammar` 등 다양한 이름으로 존재한다.

---

## 🎯 핵심 원리: **마스킹(masking)**

모델은 다음 토큰 후보를 선택할 때 "확률 분포"를 생성한다.  
그런데 **우리가 허용한 문법 범위 밖의 토큰은 아예 후보에서 제거**한다면?

→ 맞다.  
**"유효하지 않은 토큰의 확률을 0으로 만들고" → "남은 것 중에서만 선택하게" 만들면 된다.**

이걸 **logits masking** 또는 **grammar-constrained decoding** 이라고 한다.

---

## 🔒 GBNF 문법을 통한 출력 제약의 예시

예를 들어, 아래처럼 JSON 문법을 GBNF (Grammar in Backus-Naur Form)로 작성할 수 있다.

```bnf
root   ::= object
value  ::= object | array | string | number | ("true" | "false" | "null") ws

object ::=
  "{" ws (
            string ":" ws value
    ("," ws string ":" ws value)*
  )? "}" ws

string ::= "\"([^\\"\x7F\x00-\x1F]" | "\\" ["\\/bfnrt])\"" ws
```

이걸 적용하면 LLM은 반드시 `"{"`로 시작해서, `key: value` 구조를 정확히 따르며, `"`로 감싼 문자열만 허용된다는 규칙을 강제받는다.

> 즉, LLM은 여전히 확률적으로 텍스트를 생성하지만,  
> 문법적으로 **유효한 토큰만 생성할 수 있는 환경으로 "제한"**된 것이다.

---

## ✅ 예시: structured output with grammar

### 💬 입력 프롬프트
```
다음 정보를 JSON으로 변환해줘:

이름: Alice  
나이: 30  
취미: 요가
```

### 📤 출력 (Grammar 적용 후)
```json
{
  "name": "Alice",
  "age": 30,
  "hobby": "요가"
}
```

→ 이 출력은 grammar로 JSON 포맷을 강제했기 때문에 항상 유효한 JSON이 나오며,  
→ `"name"`, `"age"` 같은 key는 모델이 학습으로 기억하고 있었기 때문에 의미까지 맞는다.

---

## ⚠️ 그럼에도 중요한 한 가지: **내용은 모델이 학습해야 한다**

문법은 어디까지나 **형식적인 구조만 제어**할 수 있다.

- `"{"`, `":"`, `"}"` 등 포맷은 grammar로 제어 가능
- 하지만 `"Alice"`가 `"name"`의 값이라는 걸 아는 건 모델의 학습에 달려 있다.

즉,

> **문법은 "틀"을 잡아주고,  
> 의미 있는 "내용"은 사전 학습(pretraining or fine-tuning)에 의해 채워진다.**

---

## 💡 만약 모델이 이런 구조를 학습하지 않았다면?

### 👉 해법: 프롬프트 + 제약 기법 활용

1. **Few-shot Prompting**  
   - 예시를 여러 개 주면 모델이 구조를 따라 하기 쉬움

2. **Schema-in-Context Prompting**  
   - 출력 형식을 JSON Schema 등으로 설명해서 이해시키기

3. **Output Priming**  
   - 출력의 첫 몇 줄을 미리 제공해서 구조 유도

4. **PEFT (Parameter-Efficient Fine-Tuning)**  
   - 구조화 출력을 반복 학습시켜 행동을 조정 (예: LoRA)

---

## 🛠 실전 도구들

| 도구 | 설명 |
|------|------|
| OpenAI function calling | 함수 호출을 JSON 형태로 응답 |
| llama.cpp + grammar | 로컬 모델에 GBNF 적용 |
| jsonformer | JSON 출력용 huggingface wrapper |
| Guidance / Guardrails | 출력 구조 통제 및 검증 |
| LoRA / QLoRA | 구조적 출력을 위한 미세조정 (PEFT) |

---

## 📌 정리

| 요소 | 역할 |
|------|------|
| 확률 분포 (logits) | 모델이 다음 토큰 후보를 정함 |
| 문법 제약 (masking) | 유효한 토큰만 허용 |
| 구조화 출력 (function_call, JSON) | 응답을 parsing-friendly하게 만듦 |
| PEFT | 구조화 출력의 "행동"을 학습 |

---

## 🔚 마무리

LLM은 원래 확률적으로 동작하지만,  
**우리가 문법을 적용하고, 확률 분포를 마스킹하고, 출력 구조를 강요함으로써**,  
결과적으로 매우 정확한 structured output을 만들 수 있게 되었다.

이러한 흐름은 단지 기술적인 트릭이 아니라,  
**LLM을 프로덕션에 도입하기 위한 핵심적인 "신뢰성 확보 전략"**이다.

**Comment by james-joobs** [comment]
# Python이 GIL의 제약 아래서도 비동기 처리를 어떻게 할 수 있을까?
### — FastAPI, Uvicorn, Gunicorn, 그리고 데이터베이스까지 포함한 구조적 설명

---

## 🔒 1. Python의 GIL: 구조적 병목

GIL (Global Interpreter Lock)은 CPython 인터프리터의 메모리 안정성을 위해 존재하는 락으로, 동시에 하나의 스레드만 Python 바이트코드를 실행할 수 있도록 제한하는 역할을 합니다. (race condition 방지)

| 항목 | 설명 |
|------|------|
| GIL (Global Interpreter Lock) | CPython에서 동시에 하나의 스레드만 바이트코드를 실행할 수 있게 제한 |
| 멀티스레드 처리 | ❌ CPU 병렬화 불가 (GIL 때문) |
| I/O 중 | ✅ GIL이 자동 해제되어 다른 스레드 실행 가능 |

---

## 🔄 2. 비동기 처리의 원리: GIL을 피하지 않고, 기다림을 활용

Python의 비동기 처리(`asyncio`)는 단일 스레드 기반의 **코루틴 기반 협조형(concurrent) 처리 방식**이다.

| 구성 요소 | 설명 |
|-----------|------|
| `async def` | 비동기 함수 정의 |
| `await` | 해당 작업이 끝날 때까지 event loop에게 제어권 넘김 |
| event loop | 스케줄러 역할. 실행 가능한 코루틴들을 돌림 |
| I/O 대기 | GIL 점유 안 함. 다른 코루틴이 실행됨 |

> ✅ 비동기 처리는 GIL이 있어도 효율적인 **논리적 동시성**을 구현한다.

---

## ⚙️ 3. FastAPI + Uvicorn의 비동기 구조

### FastAPI
- ASGI 기반의 비동기 웹 프레임워크
- `async def` 기반의 API 정의 가능

### Uvicorn
- ASGI 서버
- `uvloop`(libuv 기반), `httptools` 사용
- 단일 프로세스에서 event loop를 실행
- `--workers` 옵션으로 다중 프로세스 실행 가능

```bash
uvicorn app:app --workers 4
```

---

## 🛠️ 4. Gunicorn + UvicornWorker: 병렬성과 비동기를 동시에

```bash
gunicorn app:app -k uvicorn.workers.UvicornWorker --workers 4
```

| 구성 | 설명 |
|------|------|
| Gunicorn | 멀티 프로세스 관리 (GIL 회피) |
| UvicornWorker | 각 프로세스 내에서 asyncio event loop 실행 |
| 결과 | CPU 병렬성 + I/O 비동기성 확보 |

---

## 📡 5. 데이터베이스 처리와 GIL, 비동기의 관계

### 🔍 SQLAlchemy는 I/O 바운드 작업
- 쿼리 전송 및 결과 수신은 네트워크 기반 → **I/O**
- 쿼리 생성/객체 매핑은 경량 CPU 연산

### ❌ `async def` 내에서 동기 DB 접근 (`psycopg2`) 금지
- event loop를 block하게 됨 → 병목 발생

### ✅ 비동기 ORM 필요 (SQLAlchemy Async, asyncpg 등)

```python
engine = create_async_engine("postgresql+asyncpg://...")
async with AsyncSession(engine) as session:
    result = await session.execute(select(User).where(User.id == 1))
```

---

## 🧱 6. 커넥션 풀과 세션 관리: 비동기 구조에서 중요성

| 개념 | 설명 |
|------|------|
| 커넥션 풀 | DB 연결을 재사용 → 리소스 효율 |
| 비동기 커넥션 풀 | awaitable, GIL block 없음 |
| 세션 | ORM의 논리적 단위, 커넥션을 내부적으로 소유 |
| 비동기 세션 | `async with`로 수명 명확히 관리해야 event loop block 방지 |

---

## 🧠 7. 정리: GIL 제약 속에서 비동기를 가능하게 하는 설계

| 요소 | 설계 전략 |
|------|------------|
| CPU 병렬성 | Gunicorn의 멀티 프로세스 구조로 확보 |
| I/O 비동기 | `async def` + `await` + asyncio 기반 구조 |
| DB 연동 | async ORM 사용 (`asyncpg`, `AsyncSession`) |
| 커넥션 풀 | 비동기 커넥션 풀 설정 필수 (pool_size 등) |

---

## ✅ 결론

> Python은 GIL이라는 구조적 제약을 가지고 있지만,  
> FastAPI + Uvicorn + Gunicorn + 비동기 ORM 구조를 통해  
> ✅ I/O 비동기 처리  
> ✅ 멀티코어 CPU 병렬성 확보  
> ✅ 효율적인 DB 커넥션 관리  
> 이 모두를 동시에 만족할 수 있는 **현대적인 서버 설계가 가능하다.**
